test/unit/solver/test_cuda_lu_decomposition.cpp

```diff
@@ -49,76 +50,84 @@ void check_results(
+  auto builder = CPUSparseMatrixPolicy<double>::create(10).number_of_blocks(n_grids).initial_value(1.0e-30);
   for (std::size_t i = 0; i < 10; ++i)
     for (std::size_t j = 0; j < 10; ++j)
       if (i == j || gen_bool())
         builder = builder.with_element(i, j);
 
-  SparseMatrixPolicy<double> A(builder);
+  CPUSparseMatrixPolicy<double> cpu_A(builder);
+  GPUSparseMatrixPolicy<double> gpu_A(builder);
 
   for (std::size_t i = 0; i < 10; ++i)
     for (std::size_t j = 0; j < 10; ++j)
-      if (!A.IsZero(i, j))
+      if (!cpu_A.IsZero(i, j))
         for (std::size_t i_block = 0; i_block < n_grids; ++i_block)
-          A[i_block][i][j] = get_double();
-
-  micm::CudaLuDecomposition gpu_lud(A);
-  auto gpu_LU = micm::CudaLuDecomposition::GetLUMatrices(A, 1.0e-30);
-  gpu_lud.Decompose<double, SparseMatrixPolicy>(A, gpu_LU.first, gpu_LU.second);
-  check_results<double, SparseMatrixPolicy>(
-      A, gpu_LU.first, gpu_LU.second, [&](const double a, const double b) -> void { EXPECT_NEAR(a, b, 1.0e-5); });
+        {
+          cpu_A[i_block][i][j] = get_double();
+          gpu_A[i_block][i][j] = cpu_A[i_block][i][j];
+        }
+  
+  micm::CudaLuDecomposition gpu_lud(gpu_A);
+  auto gpu_LU = micm::CudaLuDecomposition::GetLUMatrices(gpu_A, 1.0e-30);
+  gpu_A.CopyToDevice();
+  gpu_LU.first.CopyToDevice();
+  gpu_LU.second.CopyToDevice();
+  gpu_lud.Decompose<double, GPUSparseMatrixPolicy>(gpu_A, gpu_LU.first, gpu_LU.second);
+  gpu_LU.first.CopyToHost();
+  gpu_LU.second.CopyToHost();
+  check_results<double, GPUSparseMatrixPolicy>(
+      gpu_A, gpu_LU.first, gpu_LU.second, [&](const double a, const double b) -> void { EXPECT_NEAR(a, b, 1.0e-5); });
 
-  micm::LuDecomposition cpu_lud = micm::LuDecomposition::Create<double, SparseMatrixPolicy>(A);
-  auto cpu_LU = micm::LuDecomposition::GetLUMatrices<double, SparseMatrixPolicy>(A, 1.0e-30);
-  cpu_lud.Decompose<double, SparseMatrixPolicy>(A, cpu_LU.first, cpu_LU.second);
+  micm::LuDecomposition cpu_lud = micm::LuDecomposition::Create<double, CPUSparseMatrixPolicy>(cpu_A);
+  auto cpu_LU = micm::LuDecomposition::GetLUMatrices<double, CPUSparseMatrixPolicy>(cpu_A, 1.0e-30);
+  cpu_lud.Decompose<double, CPUSparseMatrixPolicy>(cpu_A, cpu_LU.first, cpu_LU.second);
 
   // checking GPU result again CPU
-  gpu_validation<double, SparseMatrixPolicy>(gpu_LU.first, cpu_LU.first, gpu_LU.second, cpu_LU.second);
+  size_t L_size = cpu_LU.first.AsVector().size();
+  size_t U_size = cpu_LU.second.AsVector().size();
+  std::vector<double> gpu_L_vector = gpu_LU.first.AsVector();
+  std::vector<double> gpu_U_vector = gpu_LU.second.AsVector();
+  std::vector<double> cpu_L_vector = cpu_LU.first.AsVector();
+  std::vector<double> cpu_U_vector = cpu_LU.second.AsVector();
+  for (int i = 0; i < L_size; ++i)
+  {
+    EXPECT_EQ(gpu_L_vector[i], cpu_L_vector[i]);
+  };
+  for (int j = 0; j < U_size; ++j)
+  {
+    EXPECT_EQ(gpu_U_vector[j], cpu_U_vector[j]);
+  };
```

Comment:
Should these be `EXPECT_DOUBLE_EQ` or `EXPECT_NEAR` since they are technically `double`s?

Reply:
Thanks. Revised as suggested.